Cython/Compiler/ExprNodes.py

```diff
@@ -13589,7 +13588,46 @@ def generate_operation_code(self, code, result_code,
                 operand1.type.unary_op('eq'),
                 operand1.result(),
                 operand2.result()))
-
+        elif operand1.type.is_struct_or_union or operand2.type.is_struct_or_union:
+            assert op in ('==', '!=')
+
+            if not operand1.type.is_struct_or_union or not operand2.type.is_struct_or_union:
+                error(self.pos, "cannot compare struct to non-struct")
+
+            if (operand1.type.is_struct_or_union and operand1.type.kind == 'union') or (operand2.type.is_struct_or_union and operand2.type.kind == 'union'):
+                error(self.pos, "cannot compare unions")
+
+            if not operand1.type.assignable_from(operand2.type):
+                error(self.pos, "cannot compare structs of different types")
+
+            self.operand1_temp = code.funcstate.allocate_temp(operand1.type, manage_ref=False)
+            self.operand2_temp = code.funcstate.allocate_temp(operand1.type, manage_ref=False)
+            code.putln(f"{self.operand1_temp} = {operand1.result()};")
+            coercion_node = operand2.coerce_to(operand1.type, code.funcstate.scope)
+            coercion_node.generate_result_code(code)
+            code.putln(f"{self.operand2_temp} = {coercion_node.result()};")
+            coercion_node.generate_disposal_code(code)
+            coercion_node.free_temps(code)
+            code.putln(f"{result_code} = {self.generate_struct_code(self.operand1_temp, self.operand2_temp, operand1.type, operand2.type, op, code)};")
+
+            code.funcstate.release_temp(self.operand1_temp)
+            code.funcstate.release_temp(self.operand2_temp)
+        elif (operand1.type.is_ctuple and operand2.type.is_ctuple) or (op in ("in", "not_in") and operand2.type.is_ctuple and any([operand1.type.assignable_from(type_) for type_ in operand2.type.components])):
+            self.operand1_temp = code.funcstate.allocate_temp(operand1.type, manage_ref=False)
+            self.operand2_temp = code.funcstate.allocate_temp(operand2.type, manage_ref=False)
+            code.putln(f"{self.operand1_temp} = {operand1.result()};")
+            if op in ("in", "not_in"):
+                code.putln(f"{self.operand2_temp} = {operand2.result()};")
+            else:
+                coercion_node = operand2.coerce_to(operand1.type, code.funcstate.scope)
```

Comment:
Coercions should not happen during code generation. They are a type-analysis thing.

Reply:
Okay, I'll work on moving this to that stage.

Reply:
Done.